### ISSUE TITLE
[feature] Make CLI reports a soft warning if query is invalid 

### ISSUE THREAD

**Issue by keith** [ast-grep/ast-grep#575]
When testing out `sg` for the first time I tried a few different queries like `class $A` for python and `if let $A = $A` for swift before I realized that the query itself has to be entirely valid syntax. I discovered this using `--debug-query` and noting that both queries have `ERROR` in their s-expressions:

```
% sg -p 'class $A' -l py --debug-query
Pattern TreeSitter (ERROR (identifier))
% sg -p 'if let $A = $A' -l swift --debug-query
Pattern TreeSitter (ERROR bound_identifier: (simple_identifier) (simple_identifier))
```

Should `sg` fail in this case to make it clear why it didn't find any matches?

**Comment by HerringtonDarkholme** [comment]
Hi Keith, thanks for trying out ast-grep!

`ERROR` is deliberately tolerated in the CLI for better ergonomics. User's input pattern may contain error because it can be a piece of code fragment.
However, users probably want to match the code regardless of the `ERROR`. As long as other AST is matched, `ERROR` can be ignored in the pattern.

ast-grep's matching algorithm is also designed to be error tolerant. Say if `$A` appears within an `ERROR`, it will be still treated as a [meta-varaible](https://github.com/ast-grep/ast-grep/commit/a19cfa714589b02ee8822b2c5c12089b19cc105c).

But I do think in your example, the `ERROR` makes no match and it makes users confused. 
I do think adding a soft error will be a good idea if no matches are found. The CLI can direct users to playground and fixes their query!

Thanks for the suggestion!

**Comment by HerringtonDarkholme** [comment]
It is shipped in 0.10.1 🎉
